### DISCUSSION TITLE
Welcome to ME3255 - Winter `25

### DISCUSSION THREAD

**Discussion by cooperrc** [cooperrc/computational-mechanics#257] [General]
Greetings, 

This is your go-to discussion forum for ME3255 - Computational Mechanics. We are going to fly through this course 
- 3 weeks
- 4 required modules
- 1 bonus module
- and lots of [Python ![Python](https://www.python.org/static/community_logos/python-powered-h-50x65.png)](https://www.python.org/)!

These discussion posts are public, so if there are any personal topics you want to discuss email/appointment is better. 

Try out some of the awesome features that can help us connect and learn Python+Computational Mechanics together, 

1. Tell us your biggest achievement in 2024
2. Add some Python code that performs a task
3. Tell us what task the code does and how it works or why you wrote it that way

_hint: use the ` ```python ``` ` backticks to make formatted Python code_

**Comment by cooperrc** [comment]
1. Tell us your biggest achievement in 2024

Assistant coaching my x2 sons little league Fall baseball team. My youngest son was _the_ youngest player and he did a great job!

2. Add some Python code that performs a task

```python
def hello_you(name):
  out = 'hello ' + name
  print(out)
  return out
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code create a function called `hello_you` that prints out `hello` and the name you enter into the input. 

e.g. 
```python
>>> hello_you('Batman');
hello Batman
```

**Comment by minhqtrinh** [comment]
1. Tell us your biggest achievement in 2024

I saw Noah Kahan perform at Fenway over the summer and it was the highlight of my year although it was not necessarily an achievement

2. Add some Python code that performs a task

```python
number = input("Enter a number:")

if int(number) % 2 == 0:
    print("The number", number,  "is even")
else:
    print("The number", number, "is odd")
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code asks the user to input a number and outputs whether the number is even or odd. It takes the input and uses the % operator to check if there is a remainder after dividing it by 2. If there is no remainder, the number is divisible by 2 and therefore an even number. However, if there is a remainder then the number is odd. These two outputs are done using an if-else statement.

**Reply by cooperrc** [reply]
Sounds like a great concert though! We can stretch out ~accomplishment~ favorite part of 2024. 

The `even-odd` checker is used for looking for checking prime numbers or finding factors of numbers. For a while, I didn't understand why a computer can't find _every_ prime number, but then I learned about integer overflow and storing numbers in a computer. With 64 bits dedicated to integers, we need one bit for plus/minus, now we have 63. The rest are powers of 2 e.g. 

```python
>>> np.iinfo(int)
iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)
```

the biggest integer that can be stored is 9,223,372,036,854,775,807. There are some tricks to prevent it from resetting to 0 after 1 is added, but we can't reliably use math and factorization with numbers this large.

**Comment by JayKay-ME** [comment]
1. Tell us your biggest achievement in 2024
     Successfully moving my family from TX to DC with a week long camping trip along the way, no issues and no lost/broken items in  transit. Or learning python on the fly in two lab classes last semester as my old university only taught MATLAB to engineers. 

2. Add some Python code that performs a task

```python 
import random
def Gambling():
   DiceFace = random.randint(1, 6)
   Gamble = int(input("Guess which side of the dice is up:  "))
   if Gamble == DiceFace:
      print("You Win!!")
   else: 
      print("(╯°□°)╯︵ ┻━┻    Get Outta Here!")
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code randomly selects "a side of a dice" or rather an integer number from 1 to 6, then prompts a user to enter a number, if the user is right they win! If the user is wrong then it prints an emoji of flipping a table over in a comedic/angry way

e.g.

```python
#call to begin:
Gambling()
Guess which side of the dice is up (1-6):  2
(╯°□°)╯︵ ┻━┻ Get Outta Here!
```

**Reply by cooperrc** [reply]
Moves are stressful, but welcome to the East Coast! DC is a great town. 

There are a few guessing games in this round of chat, that's cool. 

If it were me I would give the user at least 2 tries, but maybe I'm just a softie.

**Comment by dangallagher4** [comment]
1. Tell us your biggest achievement in 2024

I completed the second to last semester of my undergraduate education.

2. Add some Python code that performs a task

```python
import random
answer = random.randint(1, 10)
guess = int(input("Pick a number 1 through 10: "))
if guess == answer:
    print("Good guess!")
else:
    print("Incorrect, the correct number was", answer)
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code asks the user to guess a numebr 1-10, It uses the random funciton to randomly choose the correct answer, and if the user guesses correctly, it congratulates them. If they guess incorrectly, it tells them the correct guess.

**Reply by cooperrc** [reply]
Congrats on the penultimate semester completion!

This guessing game looks maddening ;). 

No 'high/low' hints?

**Comment by Natalie-Chin** [comment]
1. Tell us your biggest achievement in 2024
I was able to go to a Formula 1 race with my family this year, which was really fun as I've been watching for a couple of years.

2. Add some Python code that performs a task
```python
import numpy as np
mass = 5
damper = 10
spring_constant = 100

critical_damping = 2 * np.sqrt(spring_constant*mass)
damping_ratio = damper/critical_damping

print(damping_ratio)
```
3. Tell us what task the code does and how it works or why you wrote it that way
This code from my linear systems class last semester helps to calculate the damping ratio in a system if given the mass, damping coefficient, and spring constant. For these numbers, it gives the output of 0.22360.

**Reply by cooperrc** [reply]
Whoa, the Formula race must have been awesome. 

Nice script to determine the $\zeta$ damping ratio in 

$\ddot{x} + 2\zeta\omega \dot{x} + \omega \omega x = F(t)$

**Comment by pedro-de-rodrigues** [comment]
1.Tell us your biggest achievement in 2024

I'm trying to get my GPA up, before this semester it was below 3.0 but this Fall I got a 3.8 term GPA which brought it to be above a 3.0 GPA

2.Add some Python code that performs a task

```python
initial_time = input("Enter the time in 12hr format , just hour:time (Ex: 12:00 or 4:00) : ")
time_of_day = input("Enter the time of day (Ex: AM or PM): ")

initial_hour = int(initial_time.split(":")[0])
initial_mins = int(initial_time.split(":")[1])


if time_of_day == "AM" and 1<=initial_hour<12:
    print("Your time in 24 hr time is " + str(initial_hour) + ":" + str(initial_mins))
elif time_of_day == "PM" and initial_hour != 12:
    print("Your time in 24 hr time is " + str(int(initial_hour) + 12 ) +":"+ str(initial_mins))
elif time_of_day == "PM" and initial_hour == 12:
    print("Your time in 24 hr time is " + str(initial_hour) + ":" + str(initial_mins))
elif time_of_day == "AM" and initial_hour == 12:
    print("Your time in 24 hr time is " + "00" +":"+ str(initial_mins))
```
3.Tell us what task the code does and how it works or why you wrote it that way

The goal of this code is to convert 12 hr time to 24 time format or military time format. If the hour is greater than 12 PM and less than or equal to 11 then you add 12 and if its between 1 AM and 12 PM, the time stays the same, you just the drop AM or PM. In the case of 12 AM, it becomes 00 for the hours.

**Reply by cooperrc** [reply]
3.8 GPA in the Fall is awesome!! keep up the great work

I applaud your efforts on time conversion. [Time](https://docs.python.org/3/library/datetime.html) is surprisingly tricky to keep track of and there are lots of defined formats. 

![XKCD #2867](https://www.explainxkcd.com/wiki/images/thumb/9/9f/datetime_2x.png/679px-datetime_2x.png)

**Comment by Priya-K12** [comment]
1. Tell us your biggest achievement in 2024
 I applied to be a cadet commander of the Air Force ROTC cadet wing this past fall semester and I got it.
2. Add some Python code that performs a task
```
import random

def dad_joke():
  joke = ["I told my wife she was drawing her eyebrows too high. She looked surprised.", "I have a joke about construction, but I’m still working on it.", "Don't be sad because sad backwards is das and das not good."]
  return random.choice(joke)

print(dad_joke())
```
3. Tell us what task the code does and how it works or why you wrote it that way
 The code selects and prints a random dad joke from a list using the built in random function and the definition function.

**Reply by cooperrc** [reply]
Congratulations Cadet!! :small_airplane: 

I love a good Dad joke. Now, we could probably use the API for ChatGPT to generate random jokes too ;)

**Comment by bryanaguilar10** [comment]
1. Tell us your biggest achievement in 2024
        My biggest achievement was getting through this whole year with lots of challenges and lots of unexpected events that circulated around my personal and academic life.

2.Add some Python code that performs a task

```python
def solve_linear_equation(a, b):
  if a == 0:
   return "No solution" if b != 0 else "Infinite solutions"
  return -b / a

a = 2
b = -4
solution = solve_linear_equation(a, b)
print(f"The solution to the equation {a}x + {b} = 0 is: x = {solution}")

```
3.Tell us what task the code does and how it works or why you wrote it that way

The code solves a linear equation ax+b=0 by checking for special cases (no solution or infinite solutions) and calculating x=−b/a when 𝑎≠0. It’s written this way to handle all scenarios cleanly and efficiently.

**Reply by cooperrc** [reply]
Glad you made it through the year!

In the `return` statement, it looks like you're defining a case when slope is 0, but you'll run into some errors when you try to return the `"no solution" if ... else` because its not a defined variable or output. 

This is where `print` might be more appropriate, e.g. 

```python
if a == 0 and b == 0:
  print('infinite solutions')
  return np.nan
elif a == 0:
  print('no solutions')
  return np.nan
```

**Comment by iisabella10** [comment]
1. Tell us your biggest achievement in 2024

Apart from the many personal issues I have gone through while working and studying throughout the year, achievements that I am proud of is finally learning how to drive a stick shift car, but also becoming a seasonal preschool teacher.

2. Add some Python code that performs a task

```python
num1 = 6
num2 = 4

sum_result = num1 * num2

print("The product of", num1, "and", num2, "is:", sum_result)
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code finds the product of two predefined numbers and then displays the result. Two numbers, num1 and num2, are defined and their product is calculated by using the * operator. This number is then stored in sum_result, where the result is printed in a readable format when running the code.

**Reply by cooperrc** [reply]
I _love_ stick shift cars. I drove a manual GMC Jimmy as a teenager and a manual Honda Fit up until a year ago. I do miss using the clutch and finding gears as I drive. 

nice work on the script, if you want to get a little fancy in the `print` statement, you can use `format`

```python
print("The product of {:d} and {:d} is: {:d}".format(num1, num2, sum_result))
```

This lets you choose how the numbers are displayed more info: [Pyformat](https://pyformat.info/)

**Comment by HassEthan** [comment]
1. Tell us your biggest achievement in 2024

I went hunting for the first time this year and had an amazing meal of pheasant afterword's

2.  Add some Python code that performs a task

```python
import random

# Ask the user to input a number between 1 and 10
user_input = int(input("Enter a number between 1 and 10: "))

# Pick a random number between 1 and 10 and compare it
random_number = random.randint(1, 10)
print("The random number is:", random_number)

#Output statement based on result
if user_input == random_number:
    print("Congratulations! You guessed the correct number!")
else:
    print("Sorry, the correct number was:", random_number)
```

3. Tell us what task the code does and how it works or why you wrote it that way

This code creates a random number 1 - 10 and asks a user to predict the number. It will then inform the user if the guess was correct or incorrect based on the input and random number values.

**Reply by cooperrc** [reply]
Whoa, pheasant hunting sounds fun. That's great

This guessing game looks maddening ;).

No 'high/low' hints?

**Comment by sadracm007** [comment]
1. Tell us your biggest achievement in 2024

This fall, I successfully completed five classes while commuting 55 minutes to school Monday through Friday and working a full-time night job.

2. Add some Python code that performs a task

``` python
def  add_two_ numbers(a, b):
    return a + b
    result = add_two_numbers (5, 3)
    print ( f"the sum of 5 and 3 is {result}")
```
3. Tell us what task the code does and how it works or why you wrote it that way

this code create a function , add_two_numbers, that take 2 numbers a and b as input and returns their sum. the result from the function is stored in the variable result, which is then printed.

e.g.
```python
>>> add_two_numbers(a, b):
the sum of 5 and 3 is 8
```

**Reply by cooperrc** [reply]
5 classes + 10-hour driving/week + night shift = holy moly!

That is a lot of work to get through, definitely a huge accomplishment!

In the code, 
```python
def  add_two_ numbers(a, b):
    return a + b
    result = add_two_numbers (5, 3)
    print ( f"the sum of 5 and 3 is {result}")
```
the `result` and `print` are still inside the function statement `def` because Python uses tabs/spaces for loops+functions+if-else etc.

this should work better, 

def  add_two_numbers(a, b):
    return a + b
result = add_two_numbers (5, 3)
print ( f"the sum of 5 and 3 is {result}")

**Reply by sadracm007** [reply]
last night, I was working on HW1 and Project , but I encountered an issue with Jupyter server. Every time I tried to access it, I received a massage asking me to terminate sessions that were active. I attempted to follow the instructions to terminated the session, but the massage kept appearing, and now I am unable to connect to the Jupyter server 

It seems like I may have open too many session, causing tis issue. could you please assist me on how to resolve this problem? 

![Picture_ Jupiter](https://github.com/user-attachments/assets/90a3f190-903f-44ab-be74-3c53a18c5e60)

**Reply by sadracm007** [reply]
I submitted a ticket to report the issue, but it seems like it might take some time to get assistance.

**Reply by cooperrc** [reply]
Click the [x]-icon under **Option** to close the extra sessions. They all access your same files, so there's no need to keep all three running.

**Comment by kjy24006** [comment]
1. Tell us your biggest achievement in 2024
I started my first job and I received my first salary, which is something I am very proud of. 

2. Add some Python code that performs a task
```python
age=int(input("Please enter your age: "))
if age>=18:
   print("You are an adult")
else:
  print("You are still young")
```
3. Tell us what task the code does and how it works or why you wrote it that way
This code aims to verify if the user is an adult or he is still young based on his age. So the program prompts the user to input their age, compares it to 18 using if else statement, and finally print a message that indicates that he is an adult if his age is greater or equal to 18, else it prints a message that indicates that he is still young.

**Reply by cooperrc** [reply]
Congrats on the new job!!

The `input` function is a great way to build interactive command line programs, but its becoming a little dated. Many programs use Graphical User Interfaces (GUIs) instead. 

_But_ this might be a useful script to process a lot of user data, so it could be used as a function that checks the age and fills in a column of data for 'adult: yes/no'. 

e.g. 

```python
def adult_check(age):
  if age >= 18:
    return True
  else:
    return False
>>> adult_check(19)
True
```

**Comment by atheennaa** [comment]
1. Tell us your biggest achievement in 2024
I'm not sure if this is my biggest achievement, but I coordinated and directed a cultural performance of 30 people for UConn's Asian Nite in March earlier this year.

2. Add some Python code that performs a task
```
miles = int(input("Enter a speed in miles per hour (MPH) to convert into kilometers per minute (KPM):"))
kilo = (miles*1.609)/60
print(miles, "MPH is equal to", kilo, "KPM")
```

3. Tell us what task the code does and how it works or why you wrote it that way
This code asks the user to input a speed in units of miles per hour, which is saved as a variable named miles. It then converts it into units of kilometers per minute, saved as kilo. It prints the result as text "(miles) MPH is equal to (kilo) KPM". I wrote the code this way because it is simple and I am fairly inexperienced to coding. I am sure there is a way to write it simpler or more efficiently, I just do not know how yet (but willing to learn!)

**Comment by joeysthilaire** [comment]
1. Tell us your biggest achievement in 2024
I learned my favorite song on piano! (Interstellar: Cornfield Chase)

2. Add some Python code that performs a task
``` python
numbers = [1, 2, 3, 4, 5]
squared = [x**2 for x in numbers]
print(f"Original: {numbers}")
print(f"Squared: {squared}")
 ``` 
3. Tell us what task the code does and how it works or why you wrote it that way
This code takes a list of  ```numbers``` , squares each number in the list, and stores the results in a new list  ```squared```. It then prints both the original and squared lists.
